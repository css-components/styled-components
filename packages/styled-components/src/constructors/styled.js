// @flow
import constructWithOptions from './constructWithOptions';
import StyledComponent from '../models/StyledComponent';
import domElements from '../utils/domElements';

import type { Target } from '../types';

const styled = (tag: Target) => constructWithOptions(StyledComponent, tag);

// Shorthands for all valid HTML Elements
for (let i = 0; i < domElements.length; i += 1) {
  styled[domElements[i]] = styled(domElements[i]);
}

export default styled;
