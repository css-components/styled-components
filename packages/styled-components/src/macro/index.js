// @flow
import traverse from '@babel/traverse';
import { createMacro } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';

function styledComponentsMacro({ references, source, state, babel: { types }, config = {} }) {
  // Replace macro imports with `styled-components`.
  traverse(state.file.ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === source) {
        path.node.source.value = 'styled-components';
      }
    },
    VariableDeclaration(path) {
      path.node.declarations.forEach(declaration => {
        if (declaration.init.type === 'CallExpression') {
          const { callee, arguments: args } = declaration.init;
          if (
            callee.type === 'Identifier' &&
            callee.name === 'require' &&
            args.length === 1 &&
            args[0].value === source
          ) {
            args[0].value = 'styled-components';
          }
        }
      });
    },
  });

  // Find name of default export if it was used.
  let customImportName;
  if ('default' in references && references['default'].length > 0) {
    customImportName = references['default'][0].node.name;
  }

  // Run Babel plugin.
  const stateWithOpts = { ...state, opts: config, customImportName };
  traverse(state.file.ast, babelPlugin({ types }).visitor, undefined, stateWithOpts);

  // Don't remove the newly-modified import statement.
  return { keepImports: true };
}

export default createMacro(styledComponentsMacro, {
  configName: 'styledComponents',
});
