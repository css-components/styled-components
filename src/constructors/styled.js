// @flow
import css from './css'
import type { Interpolation, Target } from '../types'
import domElements from '../utils/domElements'
import enforceArray from '../utils/enforce-array-arguments'

export default (styledComponent: Function) => {
  const styled = (tag: Target) =>
    (strings: Array<string>, ...interpolations: Array<Interpolation>) => {
      const tagName = typeof tag === 'string' ? `.${tag}` : '(Component)'
      enforceArray(`styled${tagName}`)(strings)
      return styledComponent(tag, css(strings, ...interpolations))
    }

  // Shorthands for all valid HTML Elements
  domElements.forEach(domElement => {
    styled[domElement] = styled(domElement)
  })

  return styled
}
