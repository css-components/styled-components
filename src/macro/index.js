// @flow
import { createMacro } from 'babel-macros'
import minify from 'babel-plugin-styled-components/lib/visitors/minify'
import displayNameAndId from 'babel-plugin-styled-components/lib/visitors/displayNameAndId'
import templateLiteral from 'babel-plugin-styled-components/lib/visitors/templateLiterals'

const configName = 'styledComponents'

function styledComponentsMacro({ references, state, babel, config = {} }) {
  const t = babel.types
  // add 'import styled from styled-components'
  const literal = config.preprocess ? 'styled-components/no-parser' : 'styled-components'
  state.file.path.parent.program.body.unshift(
    t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier('styled'))],
      t.stringLiteral(literal),
    ),
  )
  // iterate over each call and modify the TaggedTemplateExpression
  references.default.forEach(referencePath => {
    const templatePath = referencePath.parentPath.parentPath
    const basePath = referencePath
    basePath.node.name = 'styled'
    const newState = {
      ...state,
      opts: config,
    }
    if (templatePath.type === 'TaggedTemplateExpression') {
      minify(templatePath, newState)
      displayNameAndId(templatePath, newState)
      templateLiteral(templatePath, newState)
    }
  })
}

module.exports = createMacro(styledComponentsMacro, { configName })
