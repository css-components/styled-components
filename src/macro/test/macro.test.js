import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-macros'
import macro from '../../macro';

pluginTester({
  plugin,
  snapshot: true,
  tests: withFilename([{
    code: `
      import styled from '../../macro'

      styled.div\`
        background: papayawhip;
      \`
    `
  },
    `
      import React from 'react';
      import macro from '../../macro'

      class Test extends React.Component {
        render() {
          return React.createElement('div', null, 'Hello World')
        }
      }
      const StyledTest = macro(Test)\`
        background: \${props => props.theme.background};
        color: red;
      \`

      export default StyledTest
    `
  ]),
})

/*
 * This adds the filename to each test so you can do require/import relative
 * to this test file.
 */
function withFilename(tests) {
  return tests.map(t => {
    const test = {babelOptions: {filename: __filename}}
    if (typeof t === 'string') {
      test.code = t
    } else {
      Object.assign(test, t)
    }
    return test
  })
}