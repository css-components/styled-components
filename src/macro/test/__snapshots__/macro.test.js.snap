// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. macros 1`] = `
"
import styled from '../../macro'

styled.div\`
  background: papayawhip;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from 'styled-components';


styled.div.withConfig({
  displayName: 'macrotest'
})(['background:papayawhip;']);
"
`;

exports[`2. macros 1`] = `
"
import React from 'react';
import macro from '../../macro'

class Test extends React.Component {
  render() {
    return React.createElement('div', null, 'Hello World')
  }
}
const StyledTest = macro(Test)\`
  background: \${props => props.theme.background};
  color: red;
\`

export default StyledTest

      ↓ ↓ ↓ ↓ ↓ ↓

import styled from 'styled-components';
import React from 'react';


class Test extends React.Component {
  render() {
    return React.createElement('div', null, 'Hello World');
  }
}
const StyledTest = styled(Test).withConfig({
  displayName: 'macrotest__StyledTest'
})(['background:', ';color:red;'], props => props.theme.background);

export default StyledTest;
"
`;
